packages/herm-js-document/modules/index.js

```diff
@@ -17,111 +17,185 @@ function ranID(value) {
-  }
+  return r;
+}
 
-  displayRange(from, to) {
-    const keys = this.getKeys();
-    let r = '';
+function deleteRange(doc, from, length) {
+  const nextDoc = duplicate(doc);
+  const keys = getKeys(doc);
+  for (let i = 0; i < length; i += 1) {
+    const k = keys[from + i];
+    nextDoc.content[k] = null;
+  }
+  return nextDoc;
+}
 
-    keys.forEach((key, index) => {
-      if (index >= from && index < to && this.content[key]) { r += this.content[key]; }
-    });
+function displayRange(doc, from, to) {
+  const keys = getKeys(doc);
+  let r = '';
 
-    return r;
-  }
+  keys.forEach((key, index) => {
+    if (index >= from && index < to && doc.content[key]) { r += doc.content[key]; }
+  });
 
-  indexOf(key) {
-    return this.keys.indexOf(key);
-  }
+  return r;
+}
 
-  merge(newDoc) {
-    const result = new HermDoc();
-    let oldIndex = 0;
-    let newIndex = 0;
+function push(doc, index, value) {
+  const nextDoc = duplicate(doc);
+  Array.from(value).forEach((e, k) => {
+    const i = ranID(e);
+    nextDoc.keys.splice(index + k, 0, i);
+    nextDoc.content[i] = e;
+  });
+  return nextDoc;
+}
 
-    while (this.keys[oldIndex] || newDoc.keys[newIndex]) {
-      if (this.keys[oldIndex] === newDoc.keys[newIndex]) {
-        result.keys.push(this.keys[oldIndex]);
+function merge(doc, newDoc) {
+  const result = {
+    keys: [],
+    content: [],
+  };
+  let oldIndex = 0;
+  let newIndex = 0;
+
+  while (doc.keys[oldIndex] || newDoc.keys[newIndex]) {
+    if (doc.keys[oldIndex] === newDoc.keys[newIndex]) {
+      result.keys.push(doc.keys[oldIndex]);
+      oldIndex += 1;
+      newIndex += 1;
+    } else {
+      while (doc.keys[oldIndex] && newDoc.keys.indexOf(doc.keys[oldIndex]) === -1) {
+        result.keys.push(doc.keys[oldIndex]);
         oldIndex += 1;
+      }
+      while (newDoc.keys[newIndex] && doc.keys.indexOf(newDoc.keys[newIndex]) === -1) {
+        result.keys.push(newDoc.keys[newIndex]);
         newIndex += 1;
-      } else {
-        while (this.keys[oldIndex] && newDoc.keys.indexOf(this.keys[oldIndex]) === -1) {
-          result.keys.push(this.keys[oldIndex]);
-          oldIndex += 1;
-        }
-        while (newDoc.keys[newIndex] && this.keys.indexOf(newDoc.keys[newIndex]) === -1) {
-          result.keys.push(newDoc.keys[newIndex]);
-          newIndex += 1;
-        }
       }
+    }
 
-      if (!this.keys[oldIndex]) {
-        result.keys = result.keys.concat(newDoc.keys.slice(newIndex));
-        break;
-      }
+    if (!doc.keys[oldIndex]) {
+      result.keys = result.keys.concat(newDoc.keys.slice(newIndex));
+      break;
+    }
 
-      if (!newDoc.keys[newIndex]) {
-        result.keys = result.keys.concat(this.keys.slice(oldIndex));
-        break;
-      }
+    if (!newDoc.keys[newIndex]) {
+      result.keys = result.keys.concat(doc.keys.slice(oldIndex));
+      break;
     }
+  }
 
-    result.content = Object.assign([], this.content);
+  result.content = Object.assign([], doc.content);
```

Comment:
What is `Object.assign([])`? You can copy array with `.slice()`, like `[].slice()`.